util/app/path/path.go

```diff
@@ -27,3 +29,53 @@ func Path(root, path string) (string, error) {
 	}
 	return appPath, nil
 }
+
+type OutOfBoundsSymlinkError struct {
+	File string
+	Err  error
+}
+
+func (e *OutOfBoundsSymlinkError) Error() string {
+	return "out of bounds symlink found"
+}
+
+// CheckOutOfBoundsSymlinks determines if basePath contains any symlinks that
+// are absolute or point to a path outside of the basePath. If found, an
+// OutOfBoundsSymlinkError is returned.
+func CheckOutOfBoundsSymlinks(basePath string) error {
+	absBasePath, err := filepath.Abs(basePath)
+	if err != nil {
+		return err
+	}
+	return filepath.Walk(absBasePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if files.IsSymlink(info) {
+			// We don't use filepath.EvalSymlinks because it fails without returning a path
+			// if the target doesn't exist.
+			target, err := os.Readlink(path)
```

Comment:
`os.Readlink` might be more clever than we want. Looks like it might actually evaluate symlinks. That's a problem if we want to avoid allowing the user to infer existence/non-existence of out-of-bounds paths. I think that might be why the e2e test is failing. We're running `os.Readlink` on `../../` and getting a permissions error on `/tmp/_argocd-repo`. I gotta read the docs some more.

Reply:
I also wonder if `Readlink` evaluates the path relative to the link file or CWD: https://man7.org/linux/man-pages/man2/readlink.2.html

Reply:
`os.Readlink` takes an absolute path and returns the exact contents of the link 'file' without following it. It shouldn't be resolving anything - that's what `filepath.EvalSymlinks` does and why I didn't use it.

I just tested it locally too - `os.Readlink` does not complain if the symlink points to a folder that has 000 filemode.